src/semantic-highlighting.ts

```diff
@@ -0,0 +1,168 @@
+        }
+      })
+    );
+  }
+
+  initialize(capabilities: ServerCapabilities) {
+    const serverCapabilities: ServerCapabilities & { semanticHighlighting?: { scopes: string[][] } } = capabilities;
+    if (!serverCapabilities.semanticHighlighting) return;
+
+    this.scopeTable = serverCapabilities.semanticHighlighting.scopes;
+  }
+
+  fillClientCapabilities(capabilities: ClientCapabilities) {
+    const textDocumentCapabilities: TextDocumentClientCapabilities & { semanticHighlightingCapabilities?: { semanticHighlighting: boolean } } = capabilities.textDocument!;
+    textDocumentCapabilities.semanticHighlightingCapabilities = {
+      semanticHighlighting: true,
+    };
+  }
+
+  async handleNotification(params: SemanticHighlightingParams) {
+    // use https://github.com/jackguo380/vim-lsp-cxx-highlight to do highlighting
+    const lines: SemanticHighlightingLine[] = params.lines.map((line) => ({ line: line.line, tokens: this.decodeTokens(line.tokens) }));
+    const symbols: any[] = [];
+    const skipped: any[] = [];
+
+    for (const line of lines) {
+      for (const token of line.tokens) {
+        symbols.push({
+          id: 0,
+          kind: token.kind,
+          ranges: [Range.create(line.line, token.character, line.line, token.character + token.length)],
+          parentKind: 'Unknown', // FIXME
+          storage: 'None', // FIXME
+        });
+        if (token.kind === 'InactiveCode') {
+          skipped.push({
+            ranges: [Range.create(line.line, token.character, line.line, token.character + token.length)],
+          });
+        }
+      }
+    }
+
+    const doc = workspace.getDocument(params.textDocument.uri);
+    await workspace.nvim.call('lsp_cxx_hl#hl#notify_symbols', [doc.bufnr, symbols]);
+    if (skipped.length) {
+      await workspace.nvim.call('lsp_cxx_hl#hl#notify_skipped', [doc.bufnr, skipped]);
+    }
+  }
+
+  // Converts a string of base64 encoded tokens into the corresponding array of SemanticHighlightingToken.
+  private decodeTokens(tokens: string): SemanticHighlightingToken[] {
+    const scopeMask = 0xffff;
+    const lenShift = 0x10;
+    const uint32Size = 4;
+    const buf = Buffer.from(tokens, 'base64');
+    const retTokens: SemanticHighlightingToken[] = [];
+    for (let i = 0, end = buf.length / uint32Size; i < end; i += 2) {
+      const start = buf.readUInt32BE(i * uint32Size);
+      const lenKind = buf.readUInt32BE((i + 1) * uint32Size);
+      const scopeIndex = lenKind & scopeMask;
+      const len = lenKind >>> lenShift;
+      const kind = this.scopeTable[scopeIndex][0];
+      retTokens.push({ character: start, scopeIndex: scopeIndex, length: len, kind: this.decodeKind(kind) });
+    }
+
+    return retTokens;
+  }
+
+  private decodeKind(kind: string) {
+    // https://github.com/llvm/llvm-project/blob/9adc7fc3cdf571bd70d5f8bda4e2e9c233c5fd63/clang-tools-extra/clangd/SemanticHighlighting.cpp#L477
```

Comment:
also provide a reference for the returned values? Or are these just vim highlightgroups?

Reply:
There are used for vim highlight groups.